### ISSUE TITLE
Open questions on p:xslt

### ISSUE THREAD

**Issue by xml-project** [xproc/3.0-steps#61]
Unless I missed something important, some questions about p:xslt seems to be open:

1. An XSLT 3.0 processor is invoked with an `initial match selection` which can be a sequence of XDM items. In the current signature of p:xslt port `source` is sequence="false". Do we stay we the item only approach or do we allow a sequence of documents to appear for port `source`? [For correction please see below]

2. An XSLT 3.0 processor may return a sequence of XDM items as primary results, but port `result` is defined as sequence="false". Do we want to change this?

3. As far as I understand our discussion yet, we still want to support XSLT 1.0 and 2.0 processors with p:xslt. If so, I think it should be a dynamic error, if the invoked processor is not 3.0 and there is not exactly one document on port `source` and this has to have an XML content-type.

**Comment by Conal-Tuohy** [comment]
Aren't the multiple output documents of an XSLT 3.0 transformation always _secondary_ results?

If so, these multiple items are not a sequence of documents, but a single document containing a sequence, and it seems to me the interpretation of the sequence should be dealt with according to the serialization specification. https://www.w3.org/TR/xslt-xquery-serialization-31/#serdm

**Comment by xml-project** [comment]
> Aren't the multiple output documents of an XSLT 3.0 transformation always secondary results?

Not as far as I understand the XSLT 3.0 specs (and the Saxon 9.8.xx) interface. It [reads](https://www.w3.org/TR/xslt-30/#invoking-initial-mode):

> and the results of processing each item are then concatenated into a single sequence, respecting the order of items in the input sequence.

The creation of a secondary result with xsl:result-document is handled in a completely different section, so they are independent to my understanding.
Did I miss something?

**Comment by Conal-Tuohy** [comment]
I was talking about the _outputs_ of the XSLT transformation, rather than the input.

As I understand it, the only way an XSLT 3 transform can produce a sequence of _documents_, is by using `xsl:result-document`; if it produces a sequence as its "principal result", then that should be output as a single document, and should not therefore be an XProc sequence, so `sequence='false'` is correct for the `result` port of `p:xslt`. That was the point I was trying to make.

**Comment by xml-project** [comment]
I think the quoted sessions talks about _output_ or what other meaning could I attach to "results of processing"? (BTW I was using 'document' in the XProc sense = An XML document, a TEXT document, a JSON document etc.)

Could you please cite any passage from the XSLT 3.0 spec to support your understanding?

I tried the following with Saxon 9.8.x:
````
<xsl:template match=".">
  <xsl:sequence select="(1, 2, 3)" />
</xsl:template>
````
The result returned is a sequence of three atomic values, hence a sequence. No result document involved.

**Comment by xml-project** [comment]
Correction of my initial post for (1): Of course in the signature of p:xslt port 'source' is sequence=true, but as the prose says, only the first item is used as initial match selection.

**Comment by Conal-Tuohy** [comment]
What I'm thinking of is "sequence normalization" as described here: https://www.w3.org/TR/xslt-xquery-serialization-31/#serdm

It is my belief that sequence normalization is a necessary step, and if that's true, then the result of your example should be a text document containing a single text node `1 2 3`. That is what I would expect the `p:xslt` step to emit on the `result` port.

**Comment by ndw** [comment]
Given the definition of "initial match selection", I think we have to allow a sequence on source. Then we have to deal with the current semantics of a sequence on that port. I guess if version>=3.0 we say that the sequence becomes the initial match selection; with version<3.0 we keep the first-value/collection semantics.

I'd prefer (I think) not to enforce sequence normalization on the results, so I guess result becomes a sequence as well.

**Comment by Conal-Tuohy** [comment]
I'm actually conflicted about this now; I can see that sequence normalization would have the effect of erasing type information in the output tree (e.g. flattening numeric types to strings, etc).

**Comment by ndw** [comment]
Yes. I wonder if we're going to need an option for this ☹️

**Comment by ndw** [comment]
Per the 18 Apri 2019 editorial meeting, we agreed to make source a sequence with the semantics Norm describes above. We agreed to make the result a sequence as well and say that if the stylesheet produces a sequence, that's what you get.

Yes, this means you will possibly get different semantics than if you run the stylesheet outside XProc. But it's simple and consistent, or so we think.

**Comment by ndw** [comment]
Per point 3: yes, we should allow users to specify 1.0 and 2.0, but it's implementation defined what processors do with the different versions.

**Comment by Conal-Tuohy** [comment]
Do I have this right? If, for example,  I have a stylesheet which converts an XML document to text using only a template that matches and copies certain text nodes, then this would produce either a single text document, or a multitude of text documents, depending on whether the XSLT is run in an XProc pipeline or not? 

If my XSLT outputs a sequence of a comment followed by an element, would they also appear in separate documents on the `result` port?

That would seem a very strange result, to me.

**Comment by Conal-Tuohy** [comment]
I've been reading up on the XSLT spec and I now see that sequence normalization should only be performed if the effective value of the `build-tree` option (specified as an attribute of `xsl:output` or `xsl:result-document`) is `true`. That is the default when `method="xml"` which means that so long as this configuration is actually respected, then when I am integrating some legacy XSLT in my XProc app I would not generally have to worry that it would output a spurious sequence. If an XSLT output declaration specified `build-tree="false"` then it could generate a sequence quite legitimately, and on that basis I'd withdraw my objection to the `p:xslt` step's `result` port being a sequence. 

NB when `method="json"` the default value of `build-tree` is false.

**Comment by ndw** [comment]
I'm a little concerned about how and where the build-tree option is exposed, but I guess that's an implementor's problem. It would appear to me that we should respect the build-tree option.

**Comment by Conal-Tuohy** [comment]
I think Saxon will expose the `build-tree` option along with other serialization options, in a SerializationProperties object, passed to this method:
http://www.saxonica.com/documentation/index.html#!javadoc/net.sf.saxon.s9api/XdmDestination@getReceiver
http://www.saxonica.com/documentation/index.html#!javadoc/net.sf.saxon.serialize/SerializationProperties

**Comment by ndw** [comment]
* Q1: “yes”
* Q2: “yes”
* Q3: “If an XSLT 1.0 or 2.0 processor is requested, it is a
       dynamic error if the input is inappropriate (e.g., a sequence
       is provided on the input port).
* Consensus: at the 10 June meeting, we convinced ourselves that
       the tree construction semantics of XSLT will produce the
       expected results when XSLT is called from XProc in the common
       case. If users explicitly select alternate serializations or
       write stylesheets that explicitly return sequences, they may
       get different results, but that’s expected even if they find it
       surprising.

**Comment by xml-project** [comment]
In order to implement and explain the way, p:xslt will work when called `version="3.0"`, I will try to spell out the consensus in some more detail:
Let us pretend a stylesheet is called in some way and now we have the "[raw result](https://www.w3.org/TR/xslt-30/#dt-raw-result)". This is basically a sequence of items.

Depending on the (implicit or explicit) value of attribute `build-tree` a "[result tree construction](https://www.w3.org/TR/xslt-30/#result-tree-construction)" will performed.

If no result tree construction is performed, the XProc step receives the raw results and the rules in ['Creating documents from XDM step results'](http://spec.xproc.org/master/head/xproc/#documents.6) apply.

If the result tree construction is performed, the XProc step receives a document. If this document has only one child and this child is a text node, a text document will appear on p:xslt's output port `result` otherwise an XML document appears on this port. 

Did I get this right?

Further question: Should we provide a way to produce a HTML document from p:xslt (e.g. by using serialization parameter `method` as differentiator?

**Comment by eriksiegel** [comment]
> Should we provide a way to produce a HTML document from p:xslt (e.g. by using serialization parameter method as differentiator?

Correct me if I'm wrong, bit my current understanding is  that the value of XSLT's `method` is stored in the document-property `serialization` and will therefore because when XProc performs the serialization...

But... hmmm... this does not have to stop the XProc processor from flagging the XSLT output as 'application/html' or whatever html content type. SO maybe the answer to your last wuestion is indeed *yes*. Not sure...

**Comment by xml-project** [comment]
@eriksiegel You are right: According to the current status of the specs it is an XML document (because 3.3 says so) with a serialization map whose 'method'-key is associated to 'html' or 'xhtml'.
Making it an HTML document does not really change much (we have no steps that accept 'html' but no 'xml'), but it seems more natural to me.

**Comment by gimsieke** [comment]
Is there anything left to discuss? 
I wouldn’t mind if the processor changes the content type of the resulting XDM document to `text/html` if the `method` map entry has the value `html`. For the `xhtml` method, maybe use `application/xhtml+xml`.
If that was the question.

**Comment by xml-project** [comment]
@gimsieke 
The question and to me more important question was: Did I spell out our consensus right.

And yes: The second question was: Do we switch to an HTML document?

**Comment by gimsieke** [comment]
> 
> 
> @gimsieke
> The question and to me more important question was: Did I spell out our consensus right.

I think so.

> 
> And yes: The second question was: Do we switch to an HTML document?

It’s only a label that we attach to the same thing according to well-defined rules. If someone doesn’t like this slight but completely determinist magic, they can change back the `content-type` property to an XML type.

**Comment by eriksiegel** [comment]
So, lets do it!

**Comment by xml-project** [comment]
@gimsieke @eriksiegel Thanks gentleman.

**Comment by ndw** [comment]
+1

**Comment by xml-project** [comment]
Just to mention a consequence of our consensus: For `xslt:result-document` with `build-tree=false` there may be documents p:xslt's port `secondary` sharing the same URI. Right?

**Comment by xml-project** [comment]
I think we can close this as it was resolved with #178 . Any objections?

**Comment by eriksiegel** [comment]
No objections

**Comment by xml-project** [comment]
Resolved with pr #178
